### PR TITLE
dev_forumtification: added backend for notification system

### PR DESCRIPTION
**Pull request by TheoHollender**
Beggining of work for the #294 issue, added backend code to calculate notifications for subscribed threads.

### CODE REVIEW

**Comment by TheoHollender** [comment]
Actual Look of the System : 
When there is no notifications :
image
Notification with one message : 
image
Notification with two :
image
Notification with three or more : 
image

**Comment by TheoHollender** [comment]
And the look of the subscribe to a thread part (might need an adjustement for the image cause i haden't any to pick)
Unsubscribe :
image
Subscribe : 
image

File: prologin/forum/migrations/0003_auto_20210226_2307.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            name='Notification',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('new_post_count', models.PositiveIntegerField(blank=True, default=0, editable=False, verbose_name='Number of posts')),
+                ('last_author', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='notification', to=settings.AUTH_USER_MODEL)),
+                ('last_post', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='notification', to='forum.Post')),
+                ('thread', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='notification', to='forum.Thread')),
+            ],
+        ),
+    ]
```
Comment (lportemo) [inline prologin/forum/migrations/0003_auto_20210226_2307.py]: Please squash your migrations

File: prologin/forum/models.py [hunk 2]
```diff
@@ -163,9 +163,27 @@ def last_post(self):
             return None
 
     def mark_read_by(self, user):
-        ReadState.objects.update_or_create(user=user,
+        robj = ReadState.objects.update_or_create(user=user,
                                            thread=self,
-                                           defaults={'post': self.last_post})
+                                           defaults={'post': self.last_post})[0]
+        
+        if robj.subtype == 1:
```
Comment (lportemo) [inline prologin/forum/models.py]: Please avoid magic numbers and use the enum

File: prologin/forum/models.py [hunk 3]
```diff
@@ -278,7 +296,43 @@ def delete(self, **kwargs):
         return super().delete(**kwargs)
 
 
-class ReadState(models.Model):
+class NotificationList(models.Model):
+    user = models.ForeignKey(settings.AUTH_USER_MODEL,
+                             related_name='notification_list',
+                             on_delete=models.CASCADE)
+
+    notification_count = models.PositiveIntegerField(verbose_name=_("Number of notifications"), editable=False, blank=True, default=0)
```
Comment (lportemo) [inline prologin/forum/models.py]: There's no need for that as you can get that count with notification_list_object.notification.count()

File: prologin/forum/models.py [hunk 4]
```diff
@@ -288,6 +342,18 @@ class ReadState(models.Model):
...
     post = models.ForeignKey(Post, related_name='read_states',
                              on_delete=models.CASCADE)
+   
+    notification = models.ForeignKey(Notification, related_name='read_states',
+                             on_delete=models.CASCADE, null=True)
+
+    subtype = EnumField(SubscribeType, db_index=True, verbose_name=_("Subscribe type"), default=SubscribeType.nothing.value)
+    def set_subtype(self, type):
+        self.subtype = type
+        self.save()
```
Comment (lportemo) [inline prologin/forum/models.py]: IMO this method is not needed

File: prologin/forum/models.py [hunk 5]
```diff
@@ -288,6 +342,18 @@ class ReadState(models.Model):
...
+
+    def set_notification(self, notif):
+        self.notification = notif
+        self.notification.save()
```
Comment (lportemo) [inline prologin/forum/models.py]: IMO this method is not needed

File: prologin/forum/models.py [hunk 6]
```diff
@@ -310,3 +376,28 @@ def thread_save_handler(sender, **kwargs):
...
+
+def last_post_all_forums():
+    return Post.visible.all().latest()
+
+def build_notification_list(user):
+
+    notification = None
+    notification_to_build = None
+    notifications = NotificationList.objects.filter(user=user)
+    print(notifications)
```
Comment (lportemo) [inline prologin/forum/models.py]: we don't want worker's stdout to be polluted with this, please remove any prints in your code and use python's logging api if needed

File: prologin/forum/templates/forum/thread_posts.html [hunk 7]
```diff
@@ -23,6 +23,19 @@
 {% block content %}
 
   <h1>{{ thread.title }}</h1>
+  {% if request.user.is_authenticated %}
+  <form style="display: inline-block" action="{% url 'forum:toggle-subscribe-post' thread.forum.slug thread.forum.pk thread.slug thread.pk %}" method="post" class="form form-inline">
```
Comment (lportemo) [inline prologin/forum/templates/forum/thread_posts.html]: please do not use inline style

File: prologin/forum/views.py [hunk 8]
```diff
@@ -13,6 +15,7 @@
 
 import forum.forms
 import forum.models
+from forum.models import ReadState,build_notification_list,Notification,NotificationList,Thread
```
Comment (lportemo) [inline prologin/forum/views.py]: Unnecessary import (see previous line)

File: prologin/forum/views.py [hunk 9]
```diff
@@ -202,7 +205,8 @@ def get_context_data(self, **kwargs):
         # query by calling paginate_queryset().
         if ((self.request.user.is_authenticated
              and not context['page_obj'].has_next())):
-            thread.mark_read_by(self.request.user)
+            robj = thread.mark_read_by(self.request.user)
+            context["subscribed"] = (robj.subtype == 1)
```
Comment (lportemo) [inline prologin/forum/views.py]: Again, please do not use magic numbers, refer to the enum

File: prologin/forum/views.py [hunk 10]
```diff
@@ -375,3 +383,68 @@ def get_success_url(self):
...
+    permission_required = 'forum.view_thread'
+    context_object_name = 'thread'
+
+    def get_success_url(self):
+        return reverse("forum:thread", kwargs=self.kwargs)
+
+    def post(self, request, *args, **kwargs):
+        thread = Thread.objects.filter(slug=self.kwargs['slug'], pk=self.kwargs['pk'])[0]
+        rs = ReadState.objects.filter(user=request.user, thread=thread)[0]
+        rs.subtype = (rs.subtype + 1) % 2
```
Comment (lportemo) [inline prologin/forum/views.py]: Sorry, we don't do that here.

File: prologin/forum/views.py [hunk 11]
```diff
@@ -375,3 +383,68 @@ def get_success_url(self):
...
+    model = forum.models.Thread
+    permission_required = 'forum.view_thread'
+    context_object_name = 'thread'
+
+    def get_success_url(self):
+        return reverse("forum:thread", kwargs=self.kwargs)
+
+    def post(self, request, *args, **kwargs):
+        thread = Thread.objects.filter(slug=self.kwargs['slug'], pk=self.kwargs['pk'])[0]
+        rs = ReadState.objects.filter(user=request.user, thread=thread)[0]
```
Comment (lportemo) [inline prologin/forum/views.py]: Please, stop doing that filter()[0] thing, in a django view 
 you can use [django.shortcuts.get_object_or_404](https://docs.djangoproject.com/fr/3.2/topics/http/shortcuts/#get-object-or-404) that will raise a 404 error if the object is not found.

File: prologin/forum/views.py [hunk 12]
```diff
@@ -375,3 +383,68 @@ def get_success_url(self):
...
+        rs.subtype = (rs.subtype + 1) % 2
+        rs.save()
+        if rs.subtype == 1:
+            messages.success(request, _("You subscribed to the thread successfully."))
+        else:
+            if rs.notification != None:
+                rs.notification.new_post_count = 0
+                rs.notification.save()
+            messages.success(request, _("You unsubscribed to the thread successfully."))
+        return redirect(self.get_success_url())
```
Comment (lportemo) [inline prologin/forum/views.py]: In the whole function you did not check if the queryset returned by filter() was empty, so again django.shortcuts.get_object_or_404 can help you save a lot of ifs

File: prologin/forum/templates/forum/notification.html [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+{% if notifications %}
+<section class="forum-listing">
+    {% for notif in notifications %}
+        <li class="{% if notif.thread.is_sticky %}notif.thread-sticky {% endif %}{% if notif.thread.is_closed %}notif.thread-closed {% endif %}{% if notif.thread.is_moved %}notif.thread-moved {% endif %}">
+            <div class="row row-vmiddle">{% spaceless %}
+            <div class="col-sm-7 col-xs-8 notif.thread-item-main">
+                <a href="{{ notif.thread.get_absolute_url }}">
+                <h3{% if not notif.thread.seen %} class="thread-unread"{% endif %}>{{ notif.thread.title }}</h3>
+                <span class="text-muted forum-meta">
+                    {% blocktrans with username=notif.thread.author.username date_full=notif.thread.date_created|date:"c" date=notif.thread.date_created|date:"DATETIME_FORMAT" %}By <span class="notif.thread-item-author">{{ username }}</span> ⋅ <span title="{{ date_full }}">{{ date }}</span>{% endblocktrans %}
```
Comment (lportemo) [inline prologin/forum/templates/forum/notification.html]: Please rework this the 3-piped date is not understandable

File: prologin/forum/views.py (hunk 12 above)
Comment (lportemo) [inline prologin/forum/views.py]: What is the template used for this view ?
If you use this view only for POST, don't use a ListView, but just a simple View and implement the post method.

**Review by lportemo — CHANGES_REQUESTED** [review]
Thanks for the PR, I think there should be a total rework of your models. They seem too complex.

For instance, I don't see the point of the NotificationList model, you may just filter notifications by user and get the len of the queryset to get the notification_count.

You are using .filter()[0] to get a single element from db, you may use instead the .get() method of the ModelManager in a try..except ObjectDoesNotExist statement or if you are in a view you may just use [django.shortcuts.get_object_or_404](https://docs.djangoproject.com/fr/3.2/topics/http/shortcuts/#get-object-or-404).

**Comment by lportemo** [comment]
I also see that you used the `@cached_property` on a method that returns a queryset. You should not do that as it may intefere with the django caching backend.

**Comment by juli0z** [comment]
Migrated to https://gitlab.com/prologin/concours/site/-/merge_requests/345

**Pull request closed without merging**